apps/condo/domains/marketplace/schema/InvoiceContext.test.js

```diff
@@ -0,0 +1,297 @@
+            })
+
+            test('support can', async () => {
+                const [o10n] = await createTestOrganization(adminClient)
+                const [objCreated] = await createTestInvoiceContext(adminClient, o10n)
+
+                const [obj, attrs] = await updateTestInvoiceContext(supportClient, objCreated.id)
+
+                expect(obj.dv).toEqual(1)
+                expect(obj.sender).toEqual(attrs.sender)
+                expect(obj.v).toEqual(2)
+                expect(obj.updatedBy).toEqual(expect.objectContaining({ id: supportClient.user.id }))
+            })
+
+            test('staff with permission can', async () => {
+                const [o10n] = await createTestOrganization(adminClient)
+                const [objCreated] = await createTestInvoiceContext(adminClient, o10n)
+
+                const client = await makeClientWithNewRegisteredAndLoggedInUser()
+                const [role] = await createTestOrganizationEmployeeRole(adminClient, o10n, { canManageInvoiceContexts: true })
+                await createTestOrganizationEmployee(adminClient, o10n, client.user, role)
+
+                const [obj, attrs] = await updateTestInvoiceContext(client, objCreated.id)
+
+                expect(obj.dv).toEqual(1)
+                expect(obj.sender).toEqual(attrs.sender)
+                expect(obj.v).toEqual(2)
+                expect(obj.updatedBy).toEqual(expect.objectContaining({ id: client.user.id }))
+            })
+
+            test('staff without permission can\'t', async () => {
+                const [o10n] = await createTestOrganization(adminClient)
+                const [objCreated] = await createTestInvoiceContext(adminClient, o10n)
+
+                const client = await makeClientWithNewRegisteredAndLoggedInUser()
+                const [role] = await createTestOrganizationEmployeeRole(adminClient, o10n, { canManageInvoiceContexts: false })
+                await createTestOrganizationEmployee(adminClient, o10n, client.user, role)
+
+                await expectToThrowAccessDeniedErrorToObj(async () => {
+                    await updateTestInvoiceContext(client, objCreated.id)
+                })
+            })
+
+            test('anonymous can\'t', async () => {
+                const [o10n] = await createTestOrganization(adminClient)
+                const [objCreated] = await createTestInvoiceContext(adminClient, o10n)
+
+                await expectToThrowAuthenticationErrorToObj(async () => {
+                    await updateTestInvoiceContext(anonymousClient, objCreated.id)
+                })
+            })
+        })
+
+        describe('hard delete', () => {
+            test('admin can\'t', async () => {
+                const [o10n] = await createTestOrganization(adminClient)
+                const [objCreated] = await createTestInvoiceContext(adminClient, o10n)
+
+                await expectToThrowAccessDeniedErrorToObj(async () => {
```

Comment:
Accessories are checked before searching the database, so you don’t need to create entities to check for deletion; you can try to delete a non-existent record